### ISSUE TITLE
Cache directory on windows should be %localappdata% by default

### ISSUE THREAD

**Issue by felicienfrancois** [electron/electron#1404]
Currently cache directory is set to %APPDATA% by default on windows
https://github.com/atom/atom-shell/blob/master/docs/api/app.md#appgetpathname

But unless what is said on the doc page, there is a path dedicated to cache on windows: %LOCALAPPDATA%

According to https://en.wikipedia.org/wiki/Environment_variable#Windows,

```
LOCALAPPDATA is the temporary files of Applications. Its uses include storing of desktop themes, Windows error reporting, caching and profiles of web browsers.
```

So cache path should be set to %LOCALAPPDATA% and HTTP Cache, GPUCache, ~~Local Storage, Cookies~~, ... should be moved to this dir.
It is important to do this because %APPDATA% is meant to be synchronised accross computer/devices unlike %LOCALAPPDATA%

**Comment by paulcbetts** [comment]
Cookies and local storage *should* roam across users

**Comment by felicienfrancois** [comment]
@paulcbetts you may be right, that's sounds the most logical thing to do.

But Chrome store all Browser data in LOCALAPPDATA. I don't know why.
Maybe a compatibility purpose or because they want to handle roam on their own.

Well in anycase, this issue is still valid for cache data (http cache and GPUCache)

**Comment by zcbenz** [comment]
Using `%LOCALAPPDATA%` for cache dir sounds good to me, but Chrome would still store cache in `%APPDATA%` as far as I know.

**Comment by felicienfrancois** [comment]
@zcbenz on windows Google Chrome store cache AND user data in `%LOCALAPPDATA%`
And more precisely in `%LOCALAPPDATA%\Google\Chrome\User Data\Default\Cache`.
Google chrome does not use `%APPDATA%`.
https://productforums.google.com/forum/#!topic/chrome/keN3Njx6gtM

__But__ `%LOCALAPPDATA%` have been introduced in windows 7. Previous version of windows does not have this environnement variables set nor the corresponding path. So %APPDATA% is a fallback path on Vista and XP.

So cache path should be set to
```javascript
process.env.LOCALAPPDATA || process.env.APPDATA
```

**Comment by jorangreef** [comment]
Dropbox recently updated their app to prefer LOCALAPPDATA over APPDATA (which roams).

Often the app data is too large to roam, i.e. it's not just cookies.

**Comment by paulcbetts** [comment]
> Previous version of windows does not have this environnement variables set nor the corresponding path

Atom Shell doesn't run on XP or Vista anyways, we don't have to worry about that

**Comment by paulcbetts** [comment]
I'm not particularly excited about this change, I think we should close it. This is causing a problem for exactly zero users right now, but moving it has the potential to break everyone currently using Atom Shell who don't write a carefully coded migration routine, and end up breaking users. Zero benefit, non-zero downside == close this bug.

**Comment by felicienfrancois** [comment]
@paulcbetts I understand your point, and I agree that this is a big concern.
The cleaner way to fix that would be to creates new API methods and deprecates old ones which allow the change without breaking old usage

I it is not worth considering, adding a notice on the documentation page would be a first step.

**Comment by felicienfrancois** [comment]
@paulcbetts Regarding the zero benefit, this is not true.
It makes it easier to migrates from another solution (like node-webkit or whatever), keeping the same policy.
Then, issues with roaming may come sooner than we expect. I don't know what is the policy of Windows 10 about roaming but it may creates issues.

**Comment by zcbenz** [comment]
Chromium doesn't actually make use of the `cache` dir for browser's cache data, all cookies and caches are stored under `appData` dir. So making `cache` dir default to `%LOCALAPPDATA%` affects nothing unless user explicitly relies on the location `cache` dir, but it also means it won't do what you wanted, e.g. move browser cache files to `%LOCALAPPDATA%`.

**Comment by felicienfrancois** [comment]
@zcbenz from my tests and from chromium documentation, Chromium uses the same paths than Chrome except it replaces `Google/Chrome` by `Chromium`: All in `%LOCALAPPDATA%`, nothing in `%APPDATA%`
https://www.chromium.org/user-experience/user-data-directory

Reminder: 
`%LOCALAPPDATA%` => `HOME/Appdata/Local/`
`%APPDATA%` => `HOME/Appdata/Roaming/`

**Comment by bux578** [comment]
> This is causing a problem for exactly zero users right now

The folder is getting bigger and bigger and in our company "Roaming" is synced on every windows login/logoff. When using plugins you get tens of thousand node_module files that slow down the syncing process.

**Comment by bux578** [comment]
> This is causing a problem for exactly zero users right now

The atom folder is getting bigger and bigger and in our company "Roaming" is synced on every windows login/logoff. When using plugins you get tens of thousand node_module files that slow down the syncing process.

**Comment by paulcbetts** [comment]
@bux578 The ~/.atom folder or node_modules isn't stored in AppData, you should file an issue about this in Atom

**Comment by zcbenz** [comment]
After a closer looking, I found that Chromium doesn't respect the `cache` directory at all in most places, for things like browser caches and GPU caches the path is hard-coded under `userData` and there is no way to change that. So I'm going to remove the `cache` path since it is useless and causes many confusions.

**Comment by etiktin** [comment]
@zcbenz, what about moving `appData ` and `userData` to `%LOCALAPPCACHE%` like in Chrome and Chromium (see  the comments above by @felicienfrancois and @jorangreef)?

I think we should change the [code](https://github.com/atom/electron/blob/01d2765e4b566df853cd583006ebc7993d48cd38/atom/browser/lib/init.coffee#L93-L94), so we will first test if the `userData` folder exists under `%APPDATA%`, if it does we won't change anything, if it doesn't (meaning it's a new install), we will use `%LOCALAPPDATA%`. That way, we won't break the experience of current users and slowly everyone will move to the new location.

Roaming should be avoided since each enterprise uses different quotas and some times they are just ridiculously low. If a developer wants the data to be synchronized between installs, he/she should use some server side mechanism and not rely on roaming (of course, they can still use `app.setPath()` to change it back, if they really need roaming).

**Comment by felicienfrancois** [comment]
+1 for @etiktin solution

**Comment by zcbenz** [comment]
I think the root problem is we don't separate users data from cache in Electron, if we keep user data like LocalStorage and Cookie in `%APPDATA%`, and then make `cache` dir work and move cache data to `% LOCALAPPDATA%`, everything should work fine?

**Comment by etiktin** [comment]
We don't know how much data each application uses. For example see #2903 where the user is talking about having more than 500 MB of data (indexedDB + WebSQL). I'm not saying that's the general case, but it can happen. `%LOCALAPPDATA%` seems like a safer default and that's probably why Chrome and Chromium use it. 

There's a similar discussion about npm [here](https://github.com/npm/npm/issues/4564) (including some examples about the issues that arise from using roaming in an enterprise environment).

**Comment by zcbenz** [comment]
`%LOCALAPPDATA%` is a suitable choice for Chrome because it has no idea of how much data the third party websites would use, and it is safe to assume websites only use localStorage for cache.

But in Electron the app author is in charge of everything, and under most cases the data written to localStorage are user preferences. Though we don't know how much data the apps would use, but the app authors do, so if they just want to put 500MB of data into `%APPDATA%` we should not prevent them.

The good thing of keeping localStorage in `%APPDATA%` is we don't need to change behavior of current apps, and app data get roaming by default.

**Comment by jorangreef** [comment]
I agree with @etiktin and his proposal would solve the issue neatly for everyone.

I know Dropbox recently moved their storage from `roaming` to `local` probably for the same reason.

Apps should only use `roaming` if they opt-in to it and know what they are doing. It should not be the default.

**Comment by zcbenz** [comment]
Another problem is Squirrel, the popular Windows installer used by Electron apps, is installing the app into `%LOCALAPPDATA%`, so if we store user data under `%LOCALAPPDATA%` we are going to face file conflicts someday.

**Comment by etiktin** [comment]
If we make the change (moving both cache and userData or just cache), we can fix the paths so conflicts won't occur. For example:
```
// Electron folders
%LOCALAPPDATA%\MyApp\Cache
%LOCALAPPDATA%\MyApp\UserData // Will hold the Local Storage, Cookies etc.

// Squirrel folders
%LOCALAPPDATA%\MyApp\app-1.0.15
%LOCALAPPDATA%\MyApp\app-1.0.16
%LOCALAPPDATA%\MyApp\Bin
...
```

**Comment by paulcbetts** [comment]
This is a huge problem for Squirrel apps if you decide to do this, because Squirrel will wipe this directory under certain circumstances (uninstall, clean install) and we will lose user data. Not to mention, this will break existing apps. 

I'm very much not excited about this change - I don't care *at all* that what we're doing now is technically wrong, and while I have some sympathy for the roaming profiles scenario, I don't think it warrants breaking software to do it.

**Comment by etiktin** [comment]
It was just an example. If that's problematic we can put it under a different path.
```
%LOCALAPPDATA%\MyApp // Squirrel
%LOCALAPPDATA%\MyAppData // Will contain cache and userData
```

Like I said in my previous [comment](https://github.com/atom/electron/issues/1404#issuecomment-145384253), it's possible to do this without breaking anything and if we choose to move both cache and userData it will be a really simple fix.

The reason to do it, is to prevent weird issues and customer complaints by enterprise customers, which are pretty much the only folks that can "enjoy" roaming to begin with. If someone really needs roaming, they can opt-in and set the path to `%APPDATA%` and by doing that they now what they are getting into.
IMO It will be easier to make the change now than later.

**Comment by paulcbetts** [comment]
> The reason to do it, is to prevent weird issues and customer complaints by enterprise customers, which are pretty much the only folks that can "enjoy" roaming to begin with

Can you provide any concrete examples? We ship the most deployed enterprise Electron Windows app in the world at Slack, and I haven't heard anyone complain about this

**Comment by felicienfrancois** [comment]
I think we're mixing up `User documents` with `App data & settings` into `User data`.
There is no reason for storing `User documents` into thoose wierd directories, most apps will propose to strore it into user  `Home` or `My documents `

So we're talking about `App data & settings`

@paulcbetts 
As a user, I expect `App data & settings` to be removed on uninstall or clean install. There are very few exception where thoose should be kept on uninstall. And in thoose rare cases, this is the responsability of the developers to explicitly store thoose important settings in roaming instead of localstorage (which is an ephemeral location by definition)
So, I don't think Squirrel wiping this dir on uninstall is a problem, this is the wanted behavior and it should be the default.

So I find the first proposition of @etiktin a good solution: it also fix the issue of most installers that does not remove cache and settings when uninstalling.

Then to avoid breaking changes, the best way is probably to test if old path exists and use it in this case.

**Comment by paulcbetts** [comment]
> As a user, I expect App data & settings to be removed on uninstall or clean install. There are very few exception where those should be kept on uninstall.

This is explicitly _not_ the policy of how Squirrel works - developers can _choose_ to do this, but Squirrel won't do it by default. You are changing that policy for Electron apps with this change.

**Comment by etiktin** [comment]
Cleaning up after uninstall, is not related to this proposal, so I rater not get into that.

Let's say I have an app called ACME and I'm using it on my company laptop. I also have a desktop which I use from time to time. I left ACME open on my laptop and started working on my PC where I also open ACME. I do a bunch of stuff there that modify userData (Local Storage, indexedDB etc.) and then I close it. The data is supposed to be synced, but I won't actually see it. Once I close ACME on my laptop the data will be rewritten so the sync disappears. When I open ACME again on the desktop, it's userData would change to the data from the laptop.

The point of this contrived story, is that roaming doesn't really work for Electron (and of course Chrome/Chromium). If you really want to sync settings/configuration you should implement that yourself. If we would have used `%LOCALAPPDATA%` the results would have been different. Each instance would have it's own data and that data would persist on that machine. `%LOCALAPPDATA%` at the very least is predictable.

**Comment by tehbilly** [comment]
As someone who's interested in using Electron in a large enterprise environment with roaming profiles, I very much would prefer `%LOCALAPPDATA%` be the default. `%APPDATA%` is defaulted to far too often, and isn't particularly friendly to the poor AD infrastructure guys.

Just because something *can* be done and *is commonly* done doesn't mean it's the best or right thing to do. For instance we have a small percentage of our workforce who are remote workers, as such they very sensitive to network bandwidth and latency. Having the default behavior be one that has even the potential for a negative experience without a compelling reason is laying the groundwork for poor user experience. Especially when the ability to revert to that behavior can be restored on an *opt-in* basis.

I think @etiktin has the right idea, not that I'm anything but another face from The Internet voicing an opinion.

> I'm very much not excited about this change - I don't care at all that what we're doing now is technically wrong, and while I have some sympathy for the roaming profiles scenario, I don't think it warrants breaking software to do it.

The beauty of this is that you have time to plan and prepare for it, which in this instance *(taking an educated guess)* should be as simple as a call to `app.setPath()`. After reading through, I'm having a hard time seeing any solid reasons against this change.

With a deprecation plan, built-in checking on startup from Electron itself, and the ability to opt-in to whatever behavior is needed should cover all use cases with sane OS-specific defaults. Exactly what I'd want to see from a platform for cross-platform development. And without breaking any current apps, to boot!

**Comment by paulcbetts** [comment]
If individual apps want to choose to put this in a different place they totally can via `app.setPath`, I don't think we need any more discussion on this, we're probably going to keep it as-is to align with what Google Chrome does

**Comment by felicienfrancois** [comment]
@paulcbetts
As [told previously](https://github.com/atom/electron/issues/1404#issuecomment-93660267), Electron has currently not the same behavior than Google Chrome. Google Chrome store user data in %LOCALAPPDATA% which is not roamed.

Aligning with Google Chrome would mean changing Electron Behavior.

Here a summary of for/against changing for %LOCALAPPDATA%:
__For:__
- This is the recommended & good way to store user data
- Prevent unexpected roaming of data (may cause issues for app developers and system admin)
- This is what Google Chrome do

__Against:__
- This is a change. Could be breaking
- May conflict with Squirrel install path

We could stat to keep the current behavior and close this topic but in this case, I think we should add a notice in the documentation about this, because this is not the expected behavior and it can cause issues.

**Comment by nxrighthere** [comment]
It will be great as an option for the developers and we will able to choose between %appdata% and %localappdata%.

**Comment by felicienfrancois** [comment]
@nxrighthere here my workaround code:
```
var app = require("app");
var path = require("path");
if (process.platform == "win32") {
    app.setPath("appData", process.env.LOCALAPPDATA);
    app.setPath("userData", path.join(process.env.LOCALAPPDATA, app.getName()));
}
```

**Comment by paulcbetts** [comment]
> It will be great as an option for the developers and we will able to choose between %appdata% and %localappdata%.

This thread is confusing people already - this is *already* an option in Electron today. The proposal is to change the default behavior

**Comment by nxrighthere** [comment]
Electron still creates the lockfile in the %appdata% instead of %localappdata% after changing the paths in the main.js.

**Comment by zcbenz** [comment]
After #2668 getting fixed, there is no bug preventing developers from use `app.setPath` to use `%localappdata%`. I'm closing this since this is a breaking change and not everyone likes it, and it is very easy to opt out of this.

**Comment by MarioHofer** [comment]
My windows home directory at work has a 1 GB restriction. Using Electron apps not changing this default behaviour may cause my machine to permanently remind me about deleting some files to get back below the 1 GB restriction.

Changing the default behaviour of Electron means, solving this issue proactive. Even if the reasons are bad programming practices by someone else. The more Electron apps there will be, the more often users will run into this mess.

**Comment by MarshallOfSound** [comment]
> My windows home directory at work has a 1 GB restriction.

Both `%APPDATA%` and `%LOCALAPPDATA%` are in your home directory?  I don't think there is anywhere else we can put these files if we wanted to.

Also you'll find that most of the space being consumed will be due to the apps using `Squirrel.Windows` which installs applications into `%LOCALAPPDATA%`.

IMO your use case is super rare and there isn't really anything as a framework we can do about it, these files have to go somewhere in your home directory by default as it is the only place we can guarantee you have write access to.  It should be up to the app to let you choose a new location for these files and use the APIs Electron provides.

**Comment by MarioHofer** [comment]
But only roaming get synced here. Although I'm not sure, how other companies handle that.

**Comment by kof** [comment]
Why is this issue closed? [this](https://github.com/spox/batali/pull/90) has been merged to develop branch, not to master…

**Comment by aabuhijleh** [comment]
> After a closer looking, I found that Chromium doesn't respect the `cache` directory at all in most places, for things like browser caches and GPU caches the path is hard-coded under `userData` and there is no way to change that. So I'm going to remove the `cache` path since it is useless and causes many confusions.

Is this still the case in more recent Electron versions?

Because I've tried to use `app.setPath('cache', newLocation)` in Electron 5.0.6 (on Windows) and it doesn't work at all. The app still uses the `userData` path.

**Comment by kalbert312** [comment]
Ran into this today. For a desktop app that might have local file system dependencies outside of the app data folders it makes more sense for `%LOCALAPPDATA%` to be the default if it's available as its basically safer. For example, my app lets a user create a project that references a local folder as a resource folder. I want to store the project file in the local app data folder to eliminate dealing with searching for project files across the system. It doesn't make sense for this project file to go to other machines in this case because the referenced directory belongs to the local system.

**Comment by atocyo** [comment]
Electron is joke. Cache data should not be stored in Roaming folder. Total disrespect to users. 2021.

**Comment by emmkimme** [comment]
My MR (Customize Electron/Chromium cache location #18309) has been closed. does it mean, I'm able to move the cache to any location ?

**Comment by pushkin-** [comment]
@emmkimme It was closed as dupe of this

**Comment by imakiro** [comment]
Could we reopen this issue because this has a _monumental_ impact on corporate backup sizes : for each user of roaming profiles on Virtual Desktops / Virtual Apps, VDI or similar technologies, those caches are backed up with the rest of the profiles because there is no way of excluding them at the source or put them someplace else.

With a hundred users, a simple retention policy of keeping just 1 week of profiles, backed up once a day, eats a terabyte of useless data, for **one** simple app such as slack.

The amount of storage space (think about the CO2 is you're sensible to that) wasted for such a simple thing as "put cache in Windows cache designated location, put user data where it belongs" is a bit boggling.

**Comment by emmkimme** [comment]
It was the purpose of this MR (Customize Electron/Chromium cache location #18309) to make the difference between cache location and user data location.

**Comment by emmkimme** [comment]
It was my assumption, but it has been confirmed as duplicate so I expect to be able to redirect the location of the cache in the next version of Electron Else I will have to adapt/redo my MR over this new version.

**Comment by nornagon** [comment]
I'm reopening this as it does seem like a reasonable request, but with the caveat that since we inherit this behavior from Chromium, it is not likely to be straightforward to change. It may come to pass that Chromium refactors in a way that makes it easier for us to make this sort of tweak possible, but without that or other in-depth work such as https://github.com/electron/electron/pull/19693, this may not be possible for a while.

**Comment by felixrieseberg** [comment]
Hey everyone! Thank you all so much for giving Electron a try and thinking about ways to make it better. As you can probably tell from how this issue has been open for ten years, so far it has never been important enough for someone to close it directly or to otherwise make the resources available for it to happen. 

If you're wondering what you can do to get this into Electron, great news - Electron is truly open source! We, the maintainers, [welcome outside contributions - especially from individuals and companies interested in joining us in maintaining this platform](https://www.electronjs.org/governance).

If you're reading this and are thinking to yourself "Alright, I'll do it!" - good for you! For smaller changes, you can usually open a well-tested and documented PR directly on this repository. For larger changes, I'd recommend that you start with a "[request for comments](https://www.electronjs.org/blog/rfcs)". This will ensure that you don't spend a lot of time on a larger change without getting feedback about your approach first. Alternatively, you could always hire a consultant, either from the maintainer community our outside, to perform the work that's important to you.

In short, I'm closing this issue because we're currently _not planning_ this work, but if you care deeply about this issue, we'd welcome your contribution.
